File: src/euler_012.hs
isqrt :: Int -> Int
isqrt = floor . sqrt . fromIntegral

factors :: Int -> [Int]
factors n = [x | x <- [1..isqrt n], n `mod` x == 0]

nfactors :: Int -> Int
nfactors n =
  sum . map (\x -> if x == n `div` x then 1 else 2) $ factors n

triangular = scanl1 (+) [1..]
answer = head . filter ((>=500) . nfactors)

main = putStrLn $ show $ answer triangular

File: src/euler_012.jl
using Base.Iterators

trigonal(n) = (n + 1) * n ÷ 2
divisors(n) = filter(x -> rem(n, x) == 0, 1:isqrt(n))
number_of_divisors(n) = sum(map(x -> (x == n / x) ? 1 : 2, divisors(n)))

function answer(lim)
    n = Iterators.map(trigonal, 1:100_000)
    m = Iterators.filter(x -> number_of_divisors(x) >= lim, n)
    Iterators.first(m)
end

answer(500) |> print
